### PR TITLE
Verify app origin in authentication requests

### PR DESCRIPTION
**Pull request by larrysalibra**
This pull request adds 2 verifications to `verifyAuthRequest` to verify that the manifest file and redirect url are on the same app origin as the one for which authentication is being requested.

It also adds a function `verifyAuthRequestAndLoadManifest` which loads the app manifest if the authentication request is valid - this is used in this commit https://github.com/blockstack/blockstack-browser/commit/d0a6a4185775e120d1e2d6a66045af91afa420fd in the browser branch https://github.com/blockstack/blockstack-browser/tree/feature/verify-app-origin to add app origin verification to the authentication process.

I've also added a utility function to check for same origin, additional test coverage and comments for some of the verification functions.

It addresses these issues: #209 #338 #339

### CODE REVIEW

File: src/auth/authVerification.js [hunk 1]
```diff
@@ -144,7 +164,37 @@ export function verifyAuthRequest(token) {
...
 }
 
-export function verifyAuthResponse(token, nameLookupURL) {
+/**
+ * Verify the authentication response is valid and
+ * fetch the app manifest file if valid. Otherwise, reject the promise.
+ * @param  {String} token the authentication request token
+ * @return {Promise} that resolves to the app manifest file in JSON format
+ * or rejects if the auth request or app manifest file is invalid
+ * @private
```
Comment (kantai) [inline src/auth/authVerification.js:173]: Should this be marked private if we're going to use this function in blockstack-browser?

**Review by kantai — COMMENTED** [review]
This looks great to me. Tested this with sign-in on an old application, with blockstack-browser@v0.21.6 linked to this version of blockstack.js and it worked seamlessly.

My only question is about how we're marking things public versus private. Should we still mark functions private if we use them in blockstack-browser?

**Comment by larrysalibra** [comment]
Thanks so much for taking a look at this @kantai !

> My only question is about how we're marking things public versus private. Should we still mark functions private if we use them in blockstack-browser?

At the moment, I have been marking as public functions that we want app developers to use in their apps. That's because the documentation script uses that tag to decide if it should show up on http://blockstack.github.io/blockstack.js/index.html - we use a bunch of functionality in the browser that wasn't really designed with the idea that it would be a public api. I'd label this in the category of technical debt.

I think we also need to make clear in the documentation which functions app developers should be considering versus functions that only make sense if you're making an alternative to the browser. What do you think?

**Comment by kantai** [comment]
> That's because the documentation script uses that tag to decide if it should show up on http://blockstack.github.io/blockstack.js/index.html - we use a bunch of functionality in the browser that wasn't really designed with the idea that it would be a public api. I'd label this in the category of technical debt.

Yeah, that seems like the best way to do this for now.

> I think we also need to make clear in the documentation which functions app developers should be considering versus functions that only make sense if you're making an alternative to the browser. What do you think?

Yeah, ultimately, we'll probably want to categorize the API, so that 'things app developers should use' can be separate from 'things to use if you want to interact with the blockstack blockchain' or 'things to use when you want to build an alternative browser' -- will look for an appropriate place to add this to a documentation issue.

**Comment by vsund** [comment]
Couldn't we simply go for making these both categories the same? In a way the Browser is just an app built with `blockstack.js`. Yes, it has some special capabilities, but after all it's an app.

I remember that this was discussed somewhere but unfortunately can't remember the result, so throwing this in here to consider :)

**Comment by kantai** [comment]
> Couldn't we simply go for making these both categories the same? In a way the Browser is just an app built with blockstack.js. Yes, it has some special capabilities, but after all it's an app.

Yeah, long-term, that's the idea, but there's different kind of apps, which is why I think it'd be a good idea to  categorize the API endpoints. Anyways, this discussion should move over to #341

**Comment by kantai** [comment]
Anyways, @larrysalibra, this looks good to merge.

**Comment by larrysalibra** [comment]
Thanks - this is merged into develop despite github not showing it as merged. Apparently deleting the branch before pushing the merge commit breaks that github functionality.

The merge commit: https://github.com/blockstack/blockstack.js/commit/6bd329553603128f58dce8b67181f0d86ba37580

**Pull request closed without merging**